compiler/internal/dce/filters.go

```diff
@@ -2,31 +2,339 @@ package dce
 
 import (
 	"go/types"
-
-	"github.com/gopherjs/gopherjs/compiler/typesutil"
+	"sort"
+	"strconv"
+	"strings"
 )
 
 // getFilters determines the DCE filters for the given object.
 // This will return an object filter and optionally return a method filter.
-func getFilters(o types.Object) (objectFilter, methodFilter string) {
-	importPath := o.Pkg().Path()
-	if typesutil.IsMethod(o) {
-		recv := typesutil.RecvType(o.Type().(*types.Signature)).Obj()
-		objectFilter = importPath + `.` + recv.Name()
-		if !o.Exported() {
-			methodFilter = importPath + `.` + o.Name() + `~`
+//
+// Typically, the object filter will always be set and the method filter
+// will be empty unless the object is an unexported method.
+// However, when the object is a method invocation on an unnamed interface type
+// the object filter will be empty and only the method filter will be set.
+// The later shouldn't happen when naming a declaration but only when creating
+// dependencies.
+func getFilters(o types.Object, tArgs []types.Type) (objectFilter, methodFilter string) {
+	if f, ok := o.(*types.Func); ok {
+		sig := f.Type().(*types.Signature)
+		if recv := sig.Recv(); recv != nil {
+			// The object is a method so the object filter is the receiver type
+			// if the receiver type is named, otherwise it's an unnamed interface.
+			typ := recv.Type()
+			if ptrType, ok := typ.(*types.Pointer); ok {
+				typ = ptrType.Elem()
+			}
+			if len(tArgs) <= 0 {
+				tArgs = getTypeArgs(typ)
+			}
+			if named, ok := typ.(*types.Named); ok {
+				objectFilter = getObjectFilter(named.Obj(), tArgs)
+			}
+
+			// The method is not exported so we only need the method filter.
+			if !o.Exported() {
+				methodFilter = getMethodFilter(o, tArgs)
+			}
+			return
 		}
-	} else {
-		objectFilter = importPath + `.` + o.Name()
 	}
+
+	// The object is not a method so we only need the object filter.
+	objectFilter = getObjectFilter(o, tArgs)
 	return
 }
 
-// getDepFilter returns the filter for the given object to be used as a dependency.
-func getDepFilter(o types.Object) string {
-	qualifiedName := o.Pkg().Path() + "." + o.Name()
-	if typesutil.IsMethod(o) {
-		qualifiedName += "~"
+// getObjectFilter returns the object filter that functions as the primary
+// name when determining if a declaration is alive or not.
+// See [naming design](./README.md#naming) for more information.
```

Comment:
Markdown won't render properly here, and elsewhere.  Consider instead:

```suggestion
// See [naming design] for more information.
//
// [naming design]: https://github.com/gopherjs/gopherjs#naming
```

Or maybe even write the new documentation in GoDoc, rather than the README, for easier linking.

Reply:
Fixed